tests/ui/cli/crate-type-flag.rs

```diff
@@ -0,0 +1,57 @@
+//! Check that `rustc`'s `--crate-type` flag accepts `--crate-type=<valid_type>` as well as the
+//! multi-value version `--crate-type=<valid_type_1>,<valid_type_2>`.
+//!
+//! This test does not try to check if the output artifacts are valid.
+
+// FIXME(#132309): add a proper `supports-crate-type` directive.
+
+// Single valid crate types should pass
+//@ revisions: lib_ rlib_ staticlib_ dylib_ cdylib_ bin_ proc_macro_
+
+//@[lib_] compile-flags: --crate-type=lib
+//@[lib_] check-pass
+
+//@[rlib_] compile-flags: --crate-type=rlib
+//@[rlib_] check-pass
+
+//@[staticlib_] compile-flags: --crate-type=staticlib
+//@[staticlib_] check-pass
+
+//@[dylib_] ignore-musl (dylibs are not supported)
+//@[dylib_] ignore-wasm (dylibs are not supported)
+//@[dylib_] compile-flags: --crate-type=dylib
+//@[dylib_] check-pass
+
+//@[cdylib_] ignore-musl (cdylibs are not supported)
+//@[cdylib_] compile-flags: --crate-type=cdylib
+//@[cdylib_] check-pass
+
+//@[bin_] compile-flags: --crate-type=bin
+//@[bin_] check-pass
+
+//@[proc_macro_] ignore-wasm (proc-macro is not supported)
+//@[proc_macro_] compile-flags: --crate-type=proc-macro
+//@[proc_macro_] check-pass
+
+//@ revisions: multivalue multivalue_combined
+
+//@[multivalue] compile-flags: --crate-type=lib,rlib,staticlib
+//@[multivalue] check-pass
+
+//@[multivalue_combined] ignore-musl (dylibs are not supported)
+//@[multivalue_combined] ignore-wasm (dylibs are not supported)
+//@[multivalue_combined] compile-flags: --crate-type=lib,rlib,staticlib --crate-type=dylib
+//@[multivalue_combined] check-pass
+
+// `proc-macro` is accepted, but `proc_macro` is not.
+//@ revisions: underscore_proc_macro
+
+//@[underscore_proc_macro] compile-flags: --crate-type=proc_macro
+//@[underscore_proc_macro] error-pattern: "unknown crate type: `proc_macro`"
```

Comment:
I intentionally used `error-pattern` because the stderr snapshots can be easy to accidentally bless away, but this forces the stderr to contain the unknown crate type error message (fuzzily matched).

Reply:
You could even use `//@ dont-check-compiler-stderr` for the 2 tests, so that the stderr snapshots are not saved thus impossible to bless away ^^.

Reply:
I believe the stderr snapshots are still compared. This is (1) useful to detect if the *exact* error message changed, but can be easily reblessed (2) let's you preview what the error message (more specifically say if we add a suggestion) will be e.g. https://github.com/rust-lang/rust/pull/134720.

Reply:
If you find this confusing, then that's https://github.com/rust-lang/rust/issues/134888  😆